src/dbnode/storage/mediator.go

```diff
@@ -192,21 +180,61 @@ func (m *mediator) Close() error {
 	return nil
 }
 
+func (m *mediator) ongoingFilesystemProcesses() {
+	log := m.opts.InstrumentOptions().Logger()
+	for {
+		select {
+		case <-m.closedCh:
+			return
+		default:
+			m.sleepFn(tickCheckInterval)
+			// See comment over mediatorTimeBarrier for an explanation of this logic.
+			mediatorTime, err := m.mediatorTimeBarrier.wait()
+			if err != nil {
+				log.Error("error within ongoingFilesystemProcesses waiting for next mediatorTime", zap.Error(err))
+			}
+
+			if err := m.RunFilesystemProcesses(noForce, mediatorTime); err != nil {
+				log.Error("error within ongoingFilesystemProcesses running filesystem processes", zap.Error(err))
+				continue
+			}
+		}
+	}
+}
+
 func (m *mediator) ongoingTick() {
+	var (
+		log          = m.opts.InstrumentOptions().Logger()
+		mediatorTime = m.nowFn()
+	)
 	for {
 		select {
 		case <-m.closedCh:
 			return
 		default:
-			// NB(xichen): if we attempt to tick while another tick
-			// is in progress, throttle a little to avoid constantly
-			// checking whether the ongoing tick is finished
-			err := m.Tick(asyncRun, noForce)
-			if err == errTickInProgress {
-				m.sleepFn(tickCheckInterval)
-			} else if err != nil {
-				log := m.opts.InstrumentOptions().Logger()
-				log.Error("error within ongoingTick", zap.Error(err))
+			m.sleepFn(tickCheckInterval)
+
+			// See comment over mediatorTimeBarrier for an explanation of this logic.
+			fsProcessWaiting := m.mediatorTimeBarrier.hasWaiter()
+			if fsProcessWaiting {
+				newMediatorTime := m.nowFn()
+				if newMediatorTime.Before(mediatorTime) {
+					instrument.EmitAndLogInvariantViolation(m.opts.InstrumentOptions(), func(l *zap.Logger) {
+						l.Error(
+							"mediator time attempted to move backwards in time",
+							zap.Time("prevTime", mediatorTime), zap.Time("newTime", mediatorTime))
+					})
+
+					// Ignore new time so we can continue looping and hopefully get one thats actually in the future
+					// next time since moving backwards in time could cause undefined behavior.
+					continue
+				} else {
+					mediatorTime = m.nowFn()
+					m.mediatorTimeBarrier.release(mediatorTime)
```

Comment:
Should the release not happen after the Tick?  Otherwise the time doesn't represent "this is a start time from which the end of a tick has been completed" as it used to (and was entrusted to).

Reply:
Yeah I actually had it the way you said before and then I removed it because we no longer rely on that guarantee in our code anywhere. The reason I introduced that restriction in the first place was because of all the weird interactions we had between the series buffer/blocks and blocks getting rotated out of the buffers and into blocks and then being flushed from there. Now that we've refactored all of the buffer logic to make those decisions based on actual state instead of just making assumptions based on time we don't need a full tick to complete before we can perform a flush (if you look you'll also see I removed all the nonsense about the shard bootstrap state before tick start as well since that is also not needed).

The only guarantee I think we need to maintain now is that a concurrent tick and filesystem processes never run with inconsistent views of the current time which is the guarantee I've made with this barrier.